### PR TITLE
add ITS sensors to parallel world

### PR DESCRIPTION
**Pull request by maciacco**
@mconcas Adding ITS sensors to parallel world to prioritize navigation of sensitive elements in case of overlapping volumes

### CODE REVIEW

File: Detectors/gconfig/g4config.in [hunk 1]
```diff
@@ -7,7 +7,7 @@
 /mcTracking/loopVerbose 1
 /mcVerbose/composedPhysicsList 2
 /mcVerbose/runAction 2   # For looping thresholds control
-#/tracking/verbose 1
+/tracking/verbose 1
```
Comment (mconcas) [inline Detectors/gconfig/g4config.in]:
```suggestion
#/tracking/verbose 1
```

Comment (mconcas) [reply Detectors/gconfig/g4config.in]: Hi, this is a remnants of the debugging session, you can remove it by accepting this change

**Review by mconcas — APPROVED** [review]

**Comment by mconcas** [comment]
> @mconcas Adding ITS sensors to parallel world to prioritize navigation of sensitive elements in case of overlapping volumes

Thanks a lot @maciacco , approving to trigger CI.

**Review by mconcas — DISMISSED** [review]

File: DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/AlignParam.h [hunk 2]
```diff
@@ -52,7 +52,7 @@ class AlignParam
   double getZ() const { return mZ; }
 
   /// apply object to geoemetry
-  bool applyToGeometry() const;
+  bool applyToGeometry(bool usePW = false, bool addSensorPW = false, bool addMetalPW = false, bool addChipPW = false) const;
```
Comment (sawenzel) [inline DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/AlignParam.h]: In principle this function has nothing to do with ITS. Yet, we pass several ITS specific arguments to it.
Could we think of a cleaner structure? Something like passing a AlignGeomOptions struct (which is then extensible)? The struct could contain ITS specific members.

File: DataFormats/Detectors/Common/src/AlignParam.cxx [hunk 3]
```diff
@@ -314,7 +328,26 @@ bool AlignParam::applyToGeometry() const
   LOG(debug) << "Aligning volume " << symname;
 
   node->Align(ginv);
-
+  std::string pathStr;
+  if (path) {
+    pathStr.append(path);
+    if (getDetFromSymName(getSymName()) == "ITS" && (pathStr.find("ITSUChip") != std::string::npos) && usePW) {
```
Comment (sawenzel) [inline DataFormats/Detectors/Common/src/AlignParam.cxx]: from a framework point of view, this looks is a bit unfortunate. We go from something generic to something that contains ITS specific code. Could we think of alternatives, in which we call detector spefific routines adjustments (callback, virtual function, etc.)?

File: Detectors/Base/include/DetectorsBase/GeometryManagerParam.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+namespace o2
+{
+
+struct GeometryManagerParam : public o2::conf::ConfigurableParamHelper<GeometryManagerParam> {
+  bool useParallelWorld = false;
+  bool addSensorToPW = false;
+  bool addMetalToPW = false;
+  bool addChipToPW = false;
+
+  O2ParamDef(GeometryManagerParam, "GeometryManagerParam");
```
Comment (sawenzel) [inline Detectors/Base/include/DetectorsBase/GeometryManagerParam.h:26]: Same thought as above. This appears to be **too** generic, while some of the keys make sense only for ITS.

Would it be possible to move this to ITS and name it something like "ITSGeom" ...

**Comment by sawenzel** [comment]
Nice progress... and sorry for my late comments. I just realized that the adjustments are not done within ITS geometry code but in the global alignment framework. Not sure if there are alternatives ... but it might make sense to shortly think about if we can do it slightly better.

**Comment by mconcas** [comment]
> Nice progress... and sorry for my late comments. I just realized that the adjustments are not done within ITS geometry code but in the global alignment framework. Not sure if there are alternatives ... but it might make sense to shortly think about if we can do it slightly better.

Now that you pointed this out, I realize that we can definitely do it in a more generic way.
We'll re-iterate on this PR. Thanks for the review.

File: DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/AlignParam.h (hunk 2 above)
Comment (maciacco) [reply DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/AlignParam.h]: This has been changed in the latest version. Now the ITS-specific configuration is handled by the ITS implementation of the o2::Detector::fillParallelWorld function

File: DataFormats/Detectors/Common/src/AlignParam.cxx (hunk 3 above)
Comment (maciacco) [reply DataFormats/Detectors/Common/src/AlignParam.cxx]: in the new implementation the volumes are added to the parallel world in a second step after the alignment. As in the other comment, the ITS code is now handling the filling of ITS volumes

File: Detectors/Base/include/DetectorsBase/GeometryManagerParam.h (hunk 4 above)
Comment (maciacco) [reply Detectors/Base/include/DetectorsBase/GeometryManagerParam.h:26]: this is now moved to a configurable inside the ITS code. The general one only includes the parameter to fill the parallel geometry

**Comment by maciacco** [comment]
Dear @sawenzel I factorized the code so that the parallel world is now filled after the alignment. This way there is no issue in the export of the geometry to file + the filling of the parallel world is handled separately by each detector (only ITS as of now) avoiding the previous clashes with the alignment framework. This required to add a new function in the base Detector class, which is overridden only for ITS. Could you please have a look to the latest version? Thank you!!

**Comment by sawenzel** [comment]
I will take a look at the latest restructuring soon. For now, I would like to suggest trying out the new improvements in ROOT, available in tag v6-32-04-alice2 of our ROOT (https://github.com/alisw/root/releases/tag/v6-32-04-alice2).

You need to enable these features like so:
```
 pw = gGeoManager->CreateParallelWorld("priority_its_sensor");
 if (getenv("TGEO_PW_USEBVH")) {
      pw->SetAccelerationMode(TGeoParallelWorld::AccelerationMode::kBVH);
    }
 if (getenv("TGEO_PW_CACHING")) {
      TGeoNavigator::SetPWSafetyCaching(true);
 }
```

I would be good to get an update on hit improvements, CPU runtime, memory. I personally found that adding just the ITS sensor gave the best improvements in hit numbers.
Once everything confirmed, we can put this into production.

**Comment by maciacco** [comment]
Dear @sawenzel thanks for the updates and for the work on the ROOT side! We will update you soon on the simulation performance using the latest version.

File: Detectors/Base/include/DetectorsBase/GeometryManagerParam.h (hunk 4 above)
Comment (sawenzel) [reply Detectors/Base/include/DetectorsBase/GeometryManagerParam.h:26]: The code is still visible here: If it is moved ... probably this class can be deleted in the commit?

File: Detectors/Base/src/Detector.cxx [hunk 5]
```diff
@@ -171,6 +171,11 @@ void Detector::addAlignableVolumes() const
   LOG(warning) << "Alignable volumes are not yet defined for " << GetName();
 }
 
+void Detector::fillParallelWorld() const
+{
+  LOG(warning) << "Parallel world is not filled for " << GetName();
+}
```
Comment (sawenzel) [inline Detectors/Base/src/Detector.cxx:176]: wouldn't issue a warning. It's not dangerous if someone doesn't add to para world. What about a simple empty implementation?

**Review by sawenzel — CHANGES_REQUESTED** [review]
Looking good to me apart from some minor cleanup requests (see inline comments)

File: Detectors/Base/include/DetectorsBase/GeometryManagerParam.h (hunk 4 above)
Comment (maciacco) [reply Detectors/Base/include/DetectorsBase/GeometryManagerParam.h:26]: I am not sure why this is not showing the updated version. In the latest commit, this class now contains only the PW flag + the flags to enable the PW optimizations, while the ITS code is in another class under the ITS directory.

File: Detectors/Base/src/Detector.cxx (hunk 5 above)
Comment (maciacco) [reply Detectors/Base/src/Detector.cxx:176]: thanks, I updated this

File: Detectors/Base/include/DetectorsBase/GeometryManagerParam.h (hunk 4 above)
Comment (sawenzel) [reply Detectors/Base/include/DetectorsBase/GeometryManagerParam.h:26]: I may have had a wrong eye here. All good for me.

**Pull request merged**